helix-core/src/main/java/org/apache/helix/controller/stages/ClusterEvent.java

```diff
@@ -114,11 +127,6 @@ public String toString() {
   }
 
   public ClusterEvent clone(String eventId) {
-    ClusterEvent newEvent = new ClusterEvent(_clusterName, _eventType, eventId);
-    newEvent.setCreationTime(_creationTime);
-    for (String attributeName : _eventAttributeMap.keySet()) {
```

Comment:
@i3wangyi I am a bit confused and not convinced by the root cause: "In ClusterEvent.java, the hashmap update itself while in its for loop".

The maps _eventAttributeMap are different objects. So addAttribute() does not add the attributes to the original event's map, but the new event's different map.

So I think the root cause could be another place (eg. another thread) updating the original event's map while in for loop.

Reply:
You're right! I'll update the root cause in #785. I was initially unaware of the different objects of `_eventAttributeMap`, cause the old code is readable at first sight and it's not thread-safe. 
This fix is still valid under multi-threads. Added one test but making it multi-threading and produce the same issue is non-trivial.

Reply:
Can you explain why the copy constructor is different from the original for-loop? Maybe I miss something, but I see that `new HashMap<>(map)` is basically also a for-loop to copy the original map entries, which means, the entries could still change and throws concurrent mod exception for the step `new HashMap<>(map)`.
Please correct me if I am wrong. Appreciate it!

Reply:
@pkuwm Thanks! I was under the impression that copy over a map is a thread-safe operation in some languages(python, e.g). I just tried it in Java, it still suffers the same exception. So I believe the right fix & less code change is to make it ConcurrentHashMap. Also added a test for it.

Reply:
Agreed that ConcurrentHashMap could protect it.